README.md
# wrtview

[![PyPI version](https://img.shields.io/pypi/v/wrtview.svg)](https://pypi.python.org/pypi/wrtview/)
[![PyPI pyversions](https://img.shields.io/pypi/pyversions/klipz.svg)](https://pypi.python.org/pypi/wrtview/)
[![MIT license](https://img.shields.io/badge/License-MIT-blue.svg)](https://github.com/ropg/wrtview/blob/master/LICENSE)
[![commits since last](https://img.shields.io/github/commits-since/ropg/wrtview/latest.svg)](https://github.com/ropg/wrtview/commits/main)

`wrtview` compactly displays data about hosts on a network centered around an OpenWRT router by combining data from various files and the output of commands on the OpenWRT. It is typically ran on a computer connected to a network that is built around an OpenWRT device, not on the OpenWRT device itself.

## Installation

wrtview is a python package on [PyPI](https://pypi.org/project/wrtview). If you have python 3.5 or newer installing is easy: 

```
python -m pip install wrtview
```

Alternatively, you can install the latest development version directly from the GitHub repository:

```
python -m pip install git+https://github.com/ropg/wrtview.git
```

## Setting up the access point for pubkey ssh access

For wrtview to work, you will need to set up ssh to provide pubkey (passwordless) access to your openwrt device. If you have a Mac or a linux machine, this is done by entering ...

```
scp ~/.ssh/id_rsa.pub root@<router>:/etc/dropbear/authorized_keys
```

... replacing `<router>` with the name or ip-address of your router. If you type your password one last time you should now be able to log into your router without a pasword by just typing `ssh root@<router>`. As soon as that works you are all set up to use wrtview.

## Using wrtview

If your router is at 192.168.1.1, all you need to do is enter `wrtview` and you'll see output like this:

```
Network 'lan' on 192.168.1.1:

P  H  192.168.1.1   openwrt          68:FF:7B:9E:9A:DC TP-LINK TECHNOLOGIES C
PADHE 192.168.1.100 MacbookPro       F0:18:98:36:06:73 Apple, Inc.             wlan1 300/240/176
PADHE 192.168.1.101 MacbookPro-wired 00:50:B6:98:C4:29 GOOD WAY IND. CO., LTD
 ADHE 192.168.1.105 iPad             26:5A:90:A8:52:73 locally administered    wlan1 24/260/188
 ADHE 192.168.1.130 JessicaPhone     EA:24:8C:29:DA:12 locally administered    wlan1 120/200/100
 AD   192.168.1.151                  76:2A:A6:21:85:EC locally administered       
PADHE 192.168.1.160 OldMacbookPro    20:C9:D0:84:02:D6 Apple, Inc.             wlan0 6/195/147
   HE 192.168.1.182 MacookAir        D0:E1:40:91:88:1E Apple, Inc.                
PADHE 192.168.1.188 iPhone-John      2E:73:1F:31:B0:1D locally administered    wlan1 740/650/550
PADHE 192.168.1.200 lights-gw        00:17:88:26:0A:26 Philips Lighting BV        
  DHE 192.168.1.201 tv               10:4F:A8:03:00:8C Sony Corporation           
PADHE 192.168.1.212 printer          3C:2A:F4:42:24:A2 Brother Industries, LT     
PAD   192.168.1.228                  2E:73:1F:31:B0:1D locally administered    wlan1 240/195/154
PADHE 192.168.1.254 switch           00:1F:28:E2:66:82 HPN Supply Chain
```

> If output goes to a terminal, wrtview will show offline hosts greyed and the router itself in bold. If the output is piped or redirected, it is just plain ASCII.

As you can see, this network has a router called 'openwrt'. By default, all the hosts on the 'lan' network on the router are displayed. In the first column you can see whether a host responded to print (`P`), is in the router's ARP table (`A`), was given a DHCP lease (`D`) and whether it is in the hosts (`H`) and ethers (`E`) files. Then there's the host's IP-address and name (the latter either from hosts, DHCP lease or ethers file). After that there's the MAC-address and the manufacturer from the vendor database.

Then if the MAC-address is found in the output of `iw <interface> station dump` for either wlan0 or wlan1, that interface is displayed with the expected throughput. You can merge in data from other access points that are serving the same network elsewere in the building, see below for details.

> As you can see the wlan0 interface is slower than wlan1, because in this case the former is on 2.4 GHz and the latter is on 5 GHz. Also, Apple handheld products randomize their MAC-addresses by default to prevent tracking, so they show up as 'locally administered', meaning they invented their own MAC-address. The Macbook Air is not on, and the stations with addresses ending in 151 and 228 have connected, but they are not in the hosts or ethers files. 

## Command line options

`--network`, `-n`

By default, wrtview will display clients in the 'lan' network on the openwrt, but you can set a different network here. To specify multiple networks simply use the -n option multiple times. Networks will be listed one after the other with a header. So the example below will list hosts on two networks: 'lan' and 'guest'.

```
wrtview -n lan -n guest 192.168.0.2
```

&nbsp;

`--wireless`, `-w`

By default, wrtview checks for clients on 'wlan0' and 'wlan1'. But you can specify any set of wireless interfaces that you would like to check for clients on. Simply use multiple times to specify multiple wireless interfaces. Each interface should be given in the format `<interface>[@<host>][:<alias>]`. The hostname part allows you to check for clients on a different OpenWRT that may be serving a different part of a building. So for instance:

```
wrtview -w wlan0:S2 -w wlan1:S5 -w wlan0@192.168.0.4:N2 -w wlan1@192.168.0.4:N5 192.168.0.2
```

will cause wrtview to connect to 192.168.0.2 for the DHCP, hosts, ethers and ARP information, and show the 'wlan0' and 'wlan1' interfaces on that system as 'S2' and 'S5' respectively. Addtionally, this will cause wrtview to connect to 192.168.0.4 for data on clients to its 'wlan0' and 'wlan1' interfaces, marking their wifi connections 'N2' and 'N5' respectively in the output.

> Note that you also need to set up passwordless ssh access like detailed above for any additional wifi access points that you want queried in this way.

&nbsp;

`--leases`, `--hosts`, `--ethers`

These options can be used to specify alternative location for the DHCP leases file, the hosts file and the ethers files. The defaults should be fine though.

&nbsp;

`--format`, `-f`

Specify (in quotes) your own output format. If not specified, wrtview uses the following default format:

```
{ping}{arp}{dhcp}{hosts}{ethers} {ip:13.13} {name:17.17} {mac:17.17} {vendor:22.22}  {wifi alias} {wifi speeds}
```

The first four fields in this format string are set to either hold a space or an A, D, H or E respectively. The fields with the numbers after them are padded with spaces and cut at the specified length so that the output lines up nicely. Apart from the fields listed in the default format, you can use `wifi` (either 'W' or ' '), `clientID`, `wifi ap host`, `wifi ap interface` and every field from the output of `iw station dump <interface>`, prepended with `wifi `.

> More information on how the format string works can be found [here](https://www.programiz.com/python-programming/methods/string/format).

&nbsp;

`--sort`, `-s`

By default, the output is sorted by IP-address. Use this option to specify any field name from above to sort on. For example:

```
wrtview -s name 192.168.0.2
```

will sort the output on hostname.

&nbsp;

`--no-ping`

wrtview will upload a small script to the router to ping entire subnets so we can inicated which hosts are alive, either by ping respinse or because they appear in the ARP table because of the ping. Use `--no-ping` if you do not want this. 

&nbsp;

`--max-ping`

Maximum number of pings executed simultaneously. Set at 254 by default.

&nbsp;

`--no-ghosts`

By default, wrtview will print any wifi stations that were found associated with probed wireless interfaces (such as 'wlan1') that were not part of any logical networks (such as 'lan') that you indicated. `--no-ghosts` suppresses this behaviour.

&nbsp;

`--no-header`

Suppresses headers like `Network 'lan' on 192.168.1.1:`

&nbsp;

`--identity`, `-i`

Use this to supply an ssh identity other than the default (which is usually `~/.ssh/id_rsa`).

&nbsp;

`--greppable`, `-g` 

This will replace all empty fields with a minus sign to make sure the number of space-separated fields in the output is the same for each line.

&nbsp;

`--version`, `-v`

Shows the version number. 

&nbsp;

## Known bugs, problems, limitations

### Running wrtview on the OpenWRT station itself

Currently wrtview is not really made to run on the OpenWRT router itself. It does try to detect if it is about to execute a command on itself and leaves off the ssh connection. But it would need python3 (which is large) installed and does things in memory that should be done differently in a memory-constrained environment. If people want this on OpenWRT it would probably be best to re-write it in Lua. Presently this is out of scope.

### IPv6

At present, wrtview does not support IPv6. It would need an additional default layout for IPv6 networks and some other changes. This is definitely on the agenda.


CHANGES.md
## 1.0.1

* More robust determination of network interface names
* Speeds now reported as rounded Mbps for rx, tx (and expected throughput if available)

## 1.0.0

* BREAKING CHANGE: networks and wifi-interfaces now with multiple `-n` and `-w` options
* Added ping script to ping entire networks from router
* Offline hosts are greyed, router itself in bold
* Fixed various bugs in presenting data
* More readable code parsing data out of command output
* More descriptive errors
* Updated vendors file for MAC->vendor lookup from IEEE with new (included) script
* Cleaner code

## 0.11.1

* Fixed bug that caused ARP-table to not read correctly
* Some minor code cleanup after pylint

## 0.11.0

* Uses `-o "BatchMode yes"` for ssh to prevent password login
* Added `--greppable` / `-g` for fixed number of fields in output
* Added a few help messages for command line options when invoked with `-h`

## 0.10.0

* Now has a `--identity` / `-i` command line option to supply an ssh identity file
* Detects if running on openwrt to execute command on, does not use ssh in that case
* More meaningful output if someting doesn't work with getting the data
* Fixed bug that yielded invalid addresses from ARP output

## 0.9.1 

* Took out f'strings' and capture_output feature of subprocess.run so we depend on Python 3.5 instead of 3.8.
* Replaced (untrue) Python 3.0 requirement by true 3.5

## 0.9.0

* Initial release


wrtview/wrtview.py
#!/usr/local/opt/python/libexec/bin/python

import os, subprocess, socket, re, ipaddress, sys, argparse, pkg_resources, traceback


def main():
    global hosts, args
    default_format = "{ping}{arp}{dhcp}{hosts}{ethers} {ip:13.13} {name:17.17} {mac:17.17} " + \
                     "{vendor:22.22}  {wifi alias} {wifi speeds}"

    # Parse command line arguments
    parser = argparse.ArgumentParser()
    parser.add_argument('--leases', default='/tmp/dhcp.leases', metavar='<leases-file>')
    parser.add_argument('--hosts', default='/etc/hosts', metavar='<hosts-file>')
    parser.add_argument('--ethers', default='/etc/ethers', metavar='<ethers-file>')
    parser.add_argument('--network', '-n', metavar='<network>', action='append',
                        help='logical network names (e.g. \'lan\') to list')
    parser.add_argument('--wireless', '-w', action='append',
                        metavar='<interface>[@<host>][:<alias>]',
                        help='wireless interfaces, see README.md')
    parser.add_argument('--format', '-f', dest='format_str', default=default_format,
                        metavar='"<format string>"')
    parser.add_argument('--sort', '-s', default='ip_as_int', metavar='<sort key>')
    parser.add_argument('--no-ghosts', action='store_true')
    parser.add_argument('--no-header', action='store_true')
    parser.add_argument('--no-ping', action='store_true',
                        help='Prevents pinging of whole network from the target OpenWRT')
    parser.add_argument('--max-ping', metavar='<#>', type=int, default=254,
                        help='Maximum number of simultaneous pings')
    parser.add_argument('--identity', '-i', metavar='<file>')
    parser.add_argument('--greppable', '-g', action='store_const', const='-', default = ' ',
                        help='fixed number of space-separated output fields')
    parser.add_argument('--version', '-v', action='version',
                        version=pkg_resources.require('wrtview')[0].version)
    parser.add_argument('router', nargs='?', metavar='<name or ip>', default='192.168.1.1')
    args = parser.parse_args()

    # argparse cannot properly have defaults of lists somehow
    if not args.wireless:
        args.wireless = ['wlan0', 'wlan1']
    if not args.network:
        args.network = ['lan']

    # get data on the network names (specified with --network / -n)
    uci_output = remote_cmd('uci show network', err="Problem reading network information")
    networks = []
    for network in args.network:
        net = {}
        for k, v in re.findall(r'^network\.' + network + '\.(.*?)=\'(.*?)\'$', uci_output,
                               re.MULTILINE):
            net[k] = v
        if not net.get('ipaddr'):
            raise Exception("Network not found: " + network)
        net['name'] = network
        net['realif']  = 'br-' + network if net.get('type') == 'bridge' else net.get('ifname', '')
        m = re.search(r'HWaddr\s+([0-9A-Fa-f:]+)', remote_cmd('ifconfig ' + net['realif'], on_err=''))
        net['mac'] = m.group(1) if m else None
        networks.append(net)

    # Print it just once, not once for each ssh command
    if args.identity and not os.path.isfile(args.identity):
        sys.stderr.write('Warning: Identity file ' + args.identity +
                         ' not accessible: No such file or directory.\n')

    # Read vendor database
    # This is the files installed in /usr/share/arp-scan when package 'arp-scan-database'
    # is installed on an openwrt box, all concatenated into the file called vendors.
    vendors = {}
    for prefix, vendor in re.findall(r'^([0-9A-F]+)\s+(.*?)$',
     pkg_resources.resource_string('wrtview', 'vendors').decode('utf-8'), re.MULTILINE):
        vendors[prefix] = vendor

    # Get the wireless station data
    stations = []
    for w in args.wireless:
        parts = w.split(":", 2)
        alias = parts[1] if len(parts) == 2 else parts[0]
        parts = parts[0].split('@', 2)
        iface = parts[0]
        whost = parts[1] if len(parts) == 2 else args.router
        ns = {}
        for line in remote_cmd('iw ' + iface + ' station dump', whost).splitlines():
            m = re.search(r'^Station ([0-9A-Fa-f:]+)', line)
            if m:
                if ns != {}:
                    stations.append(ns)
                    ns = {}
                ns['mac']= m.group(1).upper()
                ns['wifi'] = 'W'
                ns['wifi ap host'] = whost
                ns['wifi ap interface'] = iface
                ns['wifi alias'] = alias
            else:
                m = re.search(r'\s+(.*?):\s*(.*)', line)
                if m:
                    ns['wifi ' + m.group(1)] = m.group(2)
        if ns != {}:
            stations.append(ns)

    # Add 'wifi speeds': rounded values for rx, tx and (if available) expected throughput
    for s in stations:
        rx = whole_mbps(s.get('wifi rx bitrate', ''))
        tx = whole_mbps(s.get('wifi tx bitrate', ''))
        exp = whole_mbps(s.get('wifi expected throughput', ''))
        s['wifi speeds'] = rx + '/' + tx
        if exp != '':
            s['wifi speeds'] += '/' + exp

    # This will hold all hosts found by any of the methods below
    hosts = []

    # DHCP
    for line in remote_cmd('cat ' + args.leases, on_err='').splitlines():
        expire, mac, ip, name, clientID = line.split(' ')
        host = find_host('ip', ip)
        if name != "*":
            host['name'] = name
        host['mac'] = mac.upper()
        host['expire'] = int(expire)
        if clientID != "*":
            host['clientID'] = clientID
        host['dhcp'] = 'D'

    # hosts
    for ip, name in re.findall(r'^\s*([\d\.]+?)\s+([\w-]+)',
                               remote_cmd('cat ' + args.hosts, on_err=''), re.MULTILINE):
        host = find_host('ip', ip)
        host['name'] = name
        host['hosts'] = 'H'

    # Ping scan.
    if not args.no_ping:
        # upload our pingall script
        pingall = pkg_resources.resource_filename('wrtview', 'pingall')
        identity = ' -i ' + args.identity if args.identity else ''
        local_cmd('scp' + identity + ' -B ' + pingall + ' root@' + args.router + ':/tmp/pingall')
        remote_cmd('chmod a+x /tmp/pingall')
        # gather all ip numbers to be pinged
        ping_ips = []
        for net in networks:
            ping_ips += list(ipaddress.IPv4Interface(net['ipaddr'] + '/' +
                                                     net['netmask']).network.hosts())
        # ping them in batches of --max-ping
        while ping_ips:
            ping_now, ping_ips = ping_ips[:args.max_ping], ping_ips[args.max_ping:]
            ping_output = remote_cmd('/tmp/pingall ' + ' '.join([str(i) for i in ping_now]))
            for ip in ping_output.splitlines():
                find_host('ip', ip)['ping'] = 'P'
        remote_cmd('rm /tmp/pingall')

    # ARP
    for ip, mac in re.findall(r'^([\d\.]+?) dev \S+? lladdr ([0-9A-Fa-f:]+?) ',
                              remote_cmd('ip -4 neigh'), re.MULTILINE):
        host = find_host('ip', ip)
        host['mac'] = mac.upper()
        host['arp'] = 'A'

    # ethers
    for mac, name in re.findall(r'^([\dA-Fa-f:]+?)\s+([\w-]+)',
                                remote_cmd('cat ' + args.ethers, on_err=''), re.MULTILINE):
        for host in hosts:
            if host.get('mac') == mac.upper():
                host['name'] = host.get('name', '(' + name + ')')
                host['ethers'] = 'E'
            if host.get('name') == name:
                if not host.get('mac'):
                    host['mac'] = mac.upper()
                    host['mac_down'] = True
                host['ethers'] = 'E'

    # Merge in data from the wireless stations
    for s in stations:
        mac = s['mac']
        found = False
        for host in hosts:
            if mac == host.get('mac') and not host.get('mac_down'):
                found = True
                host.update(s)
        if not found:
            s['ip'] = "?"
            hosts.append(s)

    # Add info about router itself
    for network in networks:
        host = find_host('ip', network['ipaddr'])
        host['mac'] = host.get('mac', network['mac'])
        host['is_router'] = 'R'

    # Add in MAC vendor string from database
    for host in hosts:
        mac = host.get('mac')
        if mac:
            if "26AE".find(mac[1]) != -1:
                host['vendor'] = 'locally administered'
            else:
                m = mac.replace(':', '')
                vendor = vendors.get(m[0:10], vendors.get(m[0:8], vendors.get(m[0:6], "unknown")))
                host['vendor'] = vendor

    # Fix up missing keys
    for host in hosts:
        # Make sure sort key exists for all hosts
        host[args.sort] = host.get(args.sort, args.greppable)
        # make sure the fields used in format_str are ' ' (or '-' if greppable) if they didn't exist
        for field in re.findall(r'\{(.*?)[\:\}]', args.format_str):
            host[field] = host.get(field, args.greppable)
        # add numeric ip for sorting
        host['ip_as_int'] = ip2int(host.get('ip', ''))
        # determine if host is online
        if host.get('ping') == 'P' or host.get('arp') == 'A' or host.get('wifi'):
            host['online'] = 'O'

    # Sort
    sorted_hosts = sorted(hosts, key=lambda k: k[args.sort])

    # Set ANSI control sequences if we're on a tty
    if sys.stdout.isatty():
        bold   = '\u001b[1m'
        faint  = '\u001b[2m'
        normal = '\u001b[0m'
    else:
        bold = normal = faint = ''

    # Print output for each network, sorted by args.sort column
    spacer = False
    for network in networks:
        if spacer: print ('\n')
        spacer = True
        if not args.no_header:
            print ("Network '" + network['name'] + "' on " + args.router + ":\n")
        for host in sorted_hosts:
            if in_same_subnet(host.get('ip', ''), network['ipaddr'], network['netmask']):
                if not host.get('online'):
                    print (faint, end='')
                if host.get('is_router'):
                    print (bold, end='')
                print (args.format_str.format(**host) + normal)
                host['printed'] = True

    # Print the 'ghosts': wifi stations that were not found in the networks specified
    if not args.no_ghosts:
        header_printed = False
        for host in sorted_hosts:
            if host.get('wifi') and not host.get('printed'):
                if not header_printed:
                    if len(networks) == 1:
                        print ('\nWiFi Stations not in this network:\n')
                    else:
                        print ('\nWiFi Stations not in these networks:\n')
                    header_printed = True
                print (args.format_str.format(**host))


# return output from command at router as string, defaults to args.router
#
# err:    optional more human-readable string to describe what went wrong
# on_err: optional string to return instead of raising exception when error
def remote_cmd(command, router=None, err=None, on_err=False):
    global args
    if not router:
        router = args.router
    # See that router is indeed remote
    if router != socket.gethostname() and router != socket.gethostbyname(socket.gethostname()):
        identity = ' -i ' + args.identity if args.identity else ''
        command = 'ssh' + identity + ' -o "BatchMode yes" root@' + router + ' ' + command
    return local_cmd(command, err, on_err)

# Execute command locally
def local_cmd(shell_command, err=None, on_err=None):
    try:
        return subprocess.run(shell_command, shell=True, timeout=5, check=True,
                              stdout=subprocess.PIPE, stderr=subprocess.PIPE
                             ).stdout.decode('utf-8').strip()
    except Exception as e:
        if on_err:
            return on_err
        else:
            if err:
                sys.stderr.write(err + '\n\n')
            sys.stderr.write(str(e))
            if e.stderr:
                sys.stderr.write(' (' + e.stderr.strip().decode('utf-8') + ')')
            sys.stderr.write('\n')
            sys.exit(1)

# Rounds first numeric value found in string
def whole_mbps(s):
    m = re.match(r'\d+(\.\d+)?', s)
    if m:
        return str(int(float(m[0]) + 0.5))
    else:
        return ''

def ip2int(ip):
    try:
        return int(ipaddress.IPv4Address(ip))
    except:
        return 0

def in_same_subnet(ip1, ip2, mask):
    return ip2int(ip1) & ip2int(mask) == ip2int(ip2) & ip2int(mask)

# Either return existing hosts entry for k == v or create new one and return it.
def find_host(k, v, add=True):
    global hosts
    for host in hosts:
        if host[k] == v:
            return host
    if add:
        hosts.append({ k: v })
        return hosts[-1]
    return None


if __name__ == "__main__":
    main()
